scanner/src/main/java/ca/on/oicr/gsi/runscanner/scanner/processor/BaseNanoporeProcessor.java

```diff
@@ -29,4 +45,90 @@ public BaseNanoporeProcessor(Builder builder, String seqName) {
+              if (readsDirectoryForRun(path).anyMatch(Files::isDirectory)) {
+                runDirectories.add(path.toFile());
+                return FileVisitResult.SKIP_SUBTREE;
+              }
+              return FileVisitResult.CONTINUE;
+            }
+
+            @Override
+            public FileVisitResult visitFile(Path path, BasicFileAttributes basicFileAttributes)
+                throws IOException {
+              return FileVisitResult.CONTINUE;
+            }
+
+            @Override
+            public FileVisitResult visitFileFailed(Path path, IOException e) throws IOException {
+              e.printStackTrace();
+              return FileVisitResult.CONTINUE;
+            }
+
+            @Override
+            public FileVisitResult postVisitDirectory(Path path, IOException e) throws IOException {
+              return FileVisitResult.CONTINUE;
+            }
+          });
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+    return runDirectories.stream();
+  }
+
+  protected abstract Stream<Path> readsDirectoryForRun(Path path);
+
+  public NotificationDto process(File runDirectory, TimeZone tz) throws IOException {
+    final File firstFile =
+        readsDirectoryForRun(runDirectory.toPath())
+            .filter(Files::isDirectory)
+            .flatMap(
+                p -> {
+                  try (Stream<Path> files = Files.list(p)) {
+                    return files
+                        .filter(BaseNanoporeProcessor::isFileFast5)
+                        .map(Path::toFile)
+                        .findFirst()
+                        .map(Stream::of)
+                        .orElseGet(Stream::empty);
+                  } catch (IOException e) {
+                    e.printStackTrace();
+                    return Stream.empty();
+                  }
+                })
+            .findFirst()
+            .orElseThrow(
+                () -> new IOException("Cannot find FAST5 file in run directory: " + runDirectory));
+
+    NanoporeNotificationDto nnd = new NanoporeNotificationDto();
+    IHDF5StringReader reader = HDF5FactoryProvider.get().openForReading(firstFile).string();
+    nnd.setRunAlias(reader.getAttr(TRACKING_ID, "run_id"));
+    nnd.setSequencerFolderPath(runDirectory.toString());
+    nnd.setSequencerName(SEQUENCER_NAME);
+    nnd.setContainerSerialNumber(reader.getAttr(TRACKING_ID, "flow_cell_id"));
+    nnd.setContainerModel(reader.getAttr(CONTEXT_TAGS, "flowcell_type"));
+    nnd.setLaneCount(LANE_COUNT);
+    nnd.setHealthType(HealthType.UNKNOWN);
+    nnd.setStartDate(
+        ZonedDateTime.parse(reader.getAttr(TRACKING_ID, "exp_start_time")).toLocalDateTime());
+    nnd.setSoftware(
+        reader.getAttr(TRACKING_ID, "version")
+            + " + "
+            + reader.getAttr(TRACKING_ID, "protocols_version"));
+
+    return processUnique(nnd, reader);
+  }
+
+  protected abstract NanoporeNotificationDto processUnique(
```

Comment:
I don't think this method needs to return since it's going to only mutate the object. If it returned a different instance of the DTO than the one it was given, that would be weird. Also, the name isn't great. You can call it just `process` if you like (since the overload resolution makes it unique). Maybe `additionalProcess` or something like that. `Unique` isn't conveying anything about the actions of this method.